### PR TITLE
Update handling of sequence labels for plot

### LINKED ISSUE #27762: [Bug]: Inconsistent treatment of list of labels in `plot` when the input is a dataframe
### Bug summary

`plt.plot` interprets a dataframe as a sequence of series to be plotted, one per column. Labels can be specified by setting `label` to a list of labels matching the number of columns. If the dataframe has only one column, though, the `label` argument is interpreted as a single label, instead of a list with one element.

### Code for reproduction

```python
from matplotlib import pyplot as plt
import pandas as pd
import numpy as np

x = np.arange(5)
y = pd.DataFrame({'a': x, 'b': x + 1})

plt.figure('double', clear=True)
plt.plot(x, y, label=['A', 'B'])
plt.legend() # -> ok

plt.figure('single', clear=True)
plt.plot(x, y.drop('b', axis=1), label=['A'])
plt.legend() # -> legend label is "['A']"

plt.show()
```

### Actual outcome

immagine

### Expected outcome

immagine

### Additional information

_No response_

### Operating system

macOS 14.2.1

### Matplotlib Version

3.8.2

### Matplotlib Backend

MacOSX

### Python version

3.12.1 | packaged by conda-forge | (main, Dec 23 2023, 08:01:35) [Clang 16.0.6 ]

### Jupyter version

_No response_

### Installation

conda

### PR DESCRIPTION
**Pull request by rcomer**
## PR summary

Closes #27762.

* `plot(x, y, label=['foo'])` where _y_ is 1D now results in the legend label "foo" instead of "['foo']" (immediate change).
* `plot(x, y, label=sequence)` where _y_ is 1D and _sequence_ is not of length 1 is deprecated to error in future.

## PR checklist

- [x] "closes #0000" is in the body of the PR description to [link the related issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [x] new and changed code is [tested](https://matplotlib.org/devdocs/devel/testing.html)
- [n/a] *Plotting related* features are demonstrated in an [example](https://matplotlib.org/devdocs/devel/document.html#write-examples-and-tutorials)
- [X] *New Features* and *API Changes* are noted with a [directive and release note](https://matplotlib.org/devdocs/devel/api_changes.html#announce-changes-deprecations-and-new-features)
- [X] Documentation complies with [general](https://matplotlib.org/devdocs/devel/document.html#write-rest-pages) and [docstring](https://matplotlib.org/devdocs/devel/document.html#write-docstrings) guidelines

### CODE REVIEW

**Comment by rcomer** [comment]
I appear to have a `boxplot` problem :-(

**Comment by rcomer** [comment]
The second commit fixes the problem at https://github.com/matplotlib/matplotlib/pull/27767#issuecomment-1937042634.  The labels were getting passed down from `boxplot` to `bxp` in two places, and now only in the one per-dataset place.  The example now gives

image

Not really useful as a legend since all the artist handles are the same.  But perhaps less bad than the previous image :-)

File: lib/matplotlib/axes/_base.py
Comment (timhoffm) [inline lib/matplotlib/axes/_base.py:516]: For better readability I would reverse the logic:

```
if cbook.is_scalar_or_string(label):
    labels = [label] * n_datasets
else:
    ...
```

File: lib/matplotlib/tests/test_legend.py
Comment (timhoffm) [inline lib/matplotlib/tests/test_legend.py:1213]: I think checking the label on the created Line2D artist is enough. We don't have to fiddle this through the legend machinery here:

```suggestion
    line, = ax.plot([[0], [1]], label=['A'])
    assert line.get_label() == 'A'
```

**Review by timhoffm — COMMENTED** [review]
I'm unclear on the desired label behavior for boxplots.

https://matplotlib.org/stable/api/_as_gen/matplotlib.axes.Axes.bxp.html says in*bxpstats*

> label: Name of the dataset (str). If available, this will be used a tick label for the boxplot

i.e. it does not say anything about artist labels, and generally, having both tick labels and legend labels simultaneously does not make much sense. So, I don't know whether #27711 (and here as  follow up) should set the artist label at all.

Then again https://matplotlib.org/stable/api/_as_gen/matplotlib.axes.Axes.boxplot.html#matplotlib.axes.Axes.boxplot (which calls `bxp`) only says

> Labels for each dataset (one per dataset).

So, quite a standard wording for legend labels, and no word about ticks.

I suppose, we've never clearly specified what should happen.

**Comment by rcomer** [comment]
Thanks @timhoffm.  I agree with everything you say about the boxplot legends.

I can't imagine anyone would want the legend at all for the multi-dataset case since (as far as I can see) there is no way to set different styles for different data sets, and it is also not possible to pass just one label to describe multiple boxplots.  My fix above just seemed like the simplest way to get rid of the warning.

File: lib/matplotlib/axes/_base.py
Comment (rcomer) [reply lib/matplotlib/axes/_base.py]: Ah yes.  I didn't see the forest for the trees.  Done.

**Comment by timhoffm** [comment]
What we need as boxplot legends are the cases:

- Each box is colored differently and gets a legend entry; e.g.
  
  grafik

- We have multiple sets of boxes and need a label for each set

  statistische-beratung-erstellt-boxplots-mit-stata-r-und-spss

I'm not yet clear how to map these to our API. Note that we don't support multiple colors, so the first example above would need three `boxplot` calls and the second example two calls. One problem is that in the second case, we habe two dimensions of labelling, the ticks and the boxes.

Either way, putting a label on each individual box is not helpful.
- identifying multiple same-styled boxes through different labels does not make sense.
- redundantly putting the label on the ticks and the box is also questionable.

So here we could either
- only put a label on the first box of the plot. This is somewhat correct, but clashes with the sematnics of the tick labelling. I'm afraid that this API is not sustainable.
- do not label the artists at all. - I think this is pre- #27711 behavior. I'm inclined to take this step back and rethink the labelling fundamentally.

Note also, that there's a structural similarity between boxplots and bar plots in that you often habe labels on the x axis and you may have multiple colors - in the two flavors discussed above. Likely we should investigate how we can make this most similar.

**Comment by rcomer** [comment]
> do not label the artists at all. - I think this is pre- https://github.com/matplotlib/matplotlib/pull/27711 behavior. I'm inclined to take this step back and rethink the labelling fundamentally.

I have not fully followed #27711 and its issue #20512 but I _think_ some artists were getting labelled before, but in a more inconsistent way.  #27711 fixed the behaviour for when you create one boxplot per call - so that covers your first example.  The multi-boxplots-per-call basically cannot work because clearly `labels` cannot be both N tick labels and one legend label.  Tick labels and legend label should be two different parameters.

**Comment by timhoffm** [comment]
> The multi-boxplots-per-call basically cannot work because clearly labels cannot be both N tick labels and one legend label. Tick labels and legend label should be two different parameters.

This is a correct observation. We have basically two options:
1) Keep `labels` referring to the tick labels and add a `legend_labels` parameter
   *pro*: backward-compatible
   *con*: semantics of `labels` is different compared to all other plotting functions
2) Repurpose `labels` for the legend; make tick labels configurable through another parameter (either `positions` or a completely new one)
   *pro*: More in line with our usual logic
   *con*: Very strong incompatibility - would break likely most existing box plots.

**Comment by timhoffm** [comment]
@rcomer I've introduced #27780, which should resolve the boxplot issues you've had with this PR. I suggest to wait until that gets merged, then rebase on it and remove all the boxplot specifics here.

**Comment by rcomer** [comment]
Hid a bunch of comments because they discuss issues with `boxplot` that are now being handled elsewhere.  This change no longer touches `boxplot`.

**Review by timhoffm — APPROVED** [review]

File: lib/matplotlib/axes/_base.py
Comment (rcomer) [reply lib/matplotlib/axes/_base.py:520]: I'm a little sleepy but I think this `if` doesn't have an `else`.

Comment (rcomer) [reply lib/matplotlib/axes/_base.py:520]: Oh, I think I get it.  I can have a look this evening (on the wrong computer at the moment).

**Review by timhoffm — APPROVED** [review]

**Review by ksunden — APPROVED** [review]

**Comment by ksunden** [comment]
Anyone can merge on green CI

**Pull request merged**
